doc/reference/misc/data_structures.rst

```diff
@@ -0,0 +1,443 @@
+enforced by a set of rotation rules used to "fix" trees following
+modification.
+
+.. figure:: rbtree.png
+    :align: center
+    :alt: rbtree example
+    :figclass: align-center
+
+    A maximally unbalanced rbtree with a black height of two.  No more
+    nodes can be added underneath the rightmost node without
+    rebalancing.
+
+These rotations are conceptually implemented on top of a primitive
+that "swaps" the position of one node with another in the list.
+Typical implementations effect this by simply swapping the nodes
+internal "data" pointers, but because the Zephyr ``struct rbnode`` is
+intrusive, that cannot work.  Zephyr must include somewhat more
+elaborate code to handle the edge cases (for example, one swapped node
+can be the root, or the two may already be parent/child).
+
+The ``struct rbnode`` struct for a Zephyr rbtree contains only two
+pointers, representing the "left", and "right" children of a node
+within the binary tree.  Traversal of a tree for rebalancing following
+modification, however, routinely requires the ability to iterate
+"upwards" from a node as well.  It is very common for red/black trees
+in the industry to store a third "parent" pointer for this purpose.
+Zephyr avoids this requirement by building a "stack" of node pointers
+locally as it traverses downward thorugh the tree and updating it
+appropriately as modifications are made.  So a Zephyr rbtree can be
+implemented with no more runtime storage overhead than a dlist.
+
+These properties, of a balanced tree data structure that works with
+only two pointers of data per node and that works without any need for
+a memory allocation API, are quite rare in the industry and are
+somewhat unique to Zephyr.
+
+Ring Buffer
+===========
+
+For circumstances where an application needs to implement asynchronous
+"streaming" copying of data, Zephyr provides a ``struct ring_buf``
+abstraction to manage copies of such data in and our of a shared
+buffer of memory.  Ring buffers may be used in either "bytes" mode,
+where the data to be streamed is an uninterpreted array of bytes, or
+"items" mode where the data much be an integral number of 32 bit
+words.  While the underlying data structure is the same, it is not
+legal to mix these two modes on a single ring buffer instance.  A ring
+buffer initialized with a byte count must be used only with the
+"bytes" API, one initialized with a word count must use the "items"
+calls.
+
+A ``struct ring_buf`` may be placed anywhere in user-accessible
+memory, and must be initialized with ``ring_buf_init()`` before use.
+This must be provided a region of user-controlled memory for use as
+the buffer itself.  Note carefully that the units of the size of the
+buffer passed change (either bytes or words) depending on how the ring
+buffer will be used later.  Macros for combining these steps in a
+single static declaration exist for convenience.
+``RING_BUF_DECLARE()`` will declare and statically initialize a ring
```

Comment:
This would read better as a list (as we did earlier) rather than a long sentence